keps/sig-apps/4563-eviction-request-api/README.md

```diff
@@ -0,0 +1,1948 @@
+  no simple way to track this in a single condition. It is much easier to handle concurrency on a
+  standalone resource.
+- The observability of the eviction progress, interceptor reference and other useful information is
+  lacking as we can see requested in the [User Stories](#user-stories-optional).
+
+
+### Enhancing PodDisruptionBudgets
+
+The main function of today's PDBs is to count how many pods the application has via the selector.
+And how many of those pods are healthy and therefore how many can be disrupted.
+
+We would like to extend this API with the following functionality
+- Request the eviction of a pod.
+- Observe which pods are in the progress of being evicted.
+
+As a cluster admin it should be possible to request the eviction of any pod. Even the one that is
+guarded by a PDB. Currently, applications can set `maxUnavailable: 0` on a PDB and thus forbid any
+eviction. Among the reasons mentioned in the
+[EvictionRequest or Eviction subresource](#evictionrequest-or-eviction-subresource) section, it is
+also not possible to change the default eviction behavior for safety reasons. So an opt-in behavior
+would not be sufficient, as we need to solve this for all pods.
+
+We could track these requests in the PDB object itself. A PDB can track multiple pods with its
+selector. The pods do not have to be from the same application, even though they usually are. These
+pods may run in various locations and may have various reasons for their disruption. It would be
+difficult to synchronize the PDB object for applications with a large number of pods. This could
+result in inconsistent updates and conflicts as many components would race for the PDB updates.
+
+The updates to the PDB could be done solely by the eviction endpoint, but this is disqualified
+because many components override this endpoint with validating admission webhooks. And so we would
+not be sure that we are tracking all the eviction requests correctly.
+
+To decouple this, we could track the eviction requests in the Pod. Unfortunately there is also a
+race for the pod by different actors ([Pod API](#pod-api)), so the eviction request is better served
+as a standalone API that implements the synchronization.
+
+To implement better observability of the eviction, we would either have to ask each actor interested
+in the eviction to update the PDB status, which would also cause the synchronization issues.
+
+The PDB controller could also request the eviction status as it does with the scale endpoint. The
+main difference is that the scale check is done per application and not per pod. So we would run
+into scalability issues here as well. Such a pod status endpoint would be hard to enforce and also
+difficult to implement for the user.
+
+Tracking the eviction status in a pod would be feasible, but it would be hard to coordinate between
+the actors as mentioned. So it would be better to decouple the eviction request process from the pod
+as well.
```

Comment:
Have you considered improving PDB to provide more insightful status about eviction, which would allow external controllers to watch that status and react accordingly (either by scaling the app up, or doing the actions you're describing above)? This way the whole process could be tracked and synchronized outside of the core resources, and k8s would only expose sufficient information to provide that kind of automation.

Reply:
- The main problems are cardinality and synchronization, we have a set of pods guarded by a single PDB and there is a group of actors (requesters and interceptors) interested in the lifecycle of these pods. So tracking the state of the pods in the status would not scale well. An aggregate status would be less useful and would still run into update conflicts.
- High API server load: as this status would only be used on eviction, it will have outdated information. Actors interacting with a PDB will want to have the latest information about the pods (e.g. recalculating descheduling rules according to the cpu/mem utilization) and might do frequent status updates, congesting the API server.
- Not all applications use PDBs. The EvictionRequest API tries to serve as a replacement for eviction and as the main descheduling primitive. This has use cases in cluster and pod autoscaling. For example when used with HPA, we would like to downscale any application/set of pods.